Rewrite

klntsky opened the pull request.

klntsky committed:
Rewrite

klntsky committed:
use config.content_dir

limitedeternity commented on .babelrc:
```diff
@@ -1,4 +1,4 @@
 {
-  "plugins": [],
+  "plugins": ["@babel/plugin-transform-runtime"],
```
Он зачем-то нужен?

klntsky replied:
Для async/await (`regeneratorRuntime is not defined` же).

limitedeternity replied:
Используй чистые промисы, не нужно тянуть целую артиллерию, чтобы заменить async/await генераторами.

limitedeternity commented on index.html:
```diff
@@ -2,15 +2,36 @@
+        document.head.appendChild(script);
+      }
+    </script>
+    <script type="text/javascript" src="public/js/bundle.min.js"></script>
```
Относительный путь плох тем, что стоит уйти с главной страницы и до ресурса уже будет не достучаться. Так, если ты на /, то будет возвращаться файл /public/js/bundle.min.js, но если ты заходишь на /about, то браузер попытается достать /about/public/js/bundle.min.js и, разумеется, получит 404. Поддержку запуска по двойному клику нужно отбросить - слишком много подводных.

klntsky replied:
@limitedeternity 
1. у нас SPA
2. относительный путь хорош тем, что приложение работает (с ограниченной функциональностью) даже если просто открыть index.html в браузере.

> Поддержку запуска по двойному клику нужно отбросить - слишком много подводных.

А именно?

limitedeternity replied:
1) Листинг не работает
2) Нет возможности сделать path rewrite
3) Все изощрения с путями приведут к тому, что будет сложно усидеть на двух стульях, вместо разработки будем думать над тем, как расположить ресурсы так, чтобы работало и там, и там.

limitedeternity commented on package.json:
```diff
@@ -37,9 +39,18 @@
     "@babel/cli": "^7.2.3",
-    "@babel/core": "^7.3.3",
+    "@babel/core": "^7.3.4",
+    "@babel/plugin-transform-runtime": "^7.3.4",
```
transform-runtime жирнющий. Подумой.

klntsky replied:
Есть альтернативы?

limitedeternity replied:
Использовать промисы в чистом виде. Либо не транспилировать async/await.

klntsky replied:
Первое накладно, потому что потребует как минимум ещё одно переписывание.

Второе не получится.
Это должно работать в браузерах, которые работают на winxp, и, соответственно, не обновлялись несколько лет. Это не моё требование, если что (надо было раньше сказать, наверное).

limitedeternity replied:
> потому что потребует как минимум ещё одно переписывание

ЕМНИП, не потребуется. Babel просто надо по-человечески настроить.

limitedeternity commented on public/css/icons.css:
```diff
@@ -1,83 +1,95 @@
 #button-line-numbers {
-  background-image: url("/public/icons/24/list-ol.png");
+  background-image: url("../icons/24/list-ol.png");
```
Иконки лежат в /public. Такой адрес запросит /icons.

klntsky replied:
Адрес резолвится относительно css-файла. Файл в /public/css

limitedeternity commented on src/coders.js:
```diff
@@ -0,0 +1,79 @@
+const { charToByteCP866, byteToCharCP866 } = require('./cp866.js');
```
В /lib до сих пор лежит пустой cp866.js

klntsky replied:
снесу

klntsky committed:
removed lib/

limitedeternity commented on package.json:
```diff
@@ -10,14 +11,15 @@
+    "test": "ava",
     "start:dev": "node index.js",
     "start:production": "export NODE_ENV=production && node index.js",
+    "bundle:dev": "px --map-inline src/init.js public/js/bundle.min.js",
```
Что такое px?

limitedeternity replied:
> cargo install pax

Сурьезно?

limitedeternity commented:
При загрузке мобильного лейаута случается следующее:

![2019-03-02 14 32 29](https://user-images.githubusercontent.com/24318966/53681266-11d15c00-3cf8-11e9-9e66-a4d828af535d.png)

И в мобильном лейауте есть один неприятный баг: когда жмешь на кнопку "меню", сразу же открывается листинг директории. Приходится нажимать кнопку "закрыть", чтобы перейти в меню.

limitedeternity committed:
Sync changes

limitedeternity commented:
Также, вероятно, `README.md` нужно обновить.

klntsky commented:
> При загрузке мобильного лейаута случается следующее:

Это баг fastclick, который пока не важен.

> И в мобильном лейауте есть один неприятный баг: когда жмешь на кнопку "меню", сразу же открывается листинг директории. Приходится нажимать кнопку "закрыть", чтобы перейти в меню.

Это не баг. Я специально это добавил задолго до переписывания. Это нужно потому, что на мобильных практически никто не пользуется списком локальных файлов, а листинг вызывается гораздо чаще. И ЕМНИП это должно срабатывать только если локальных файлов нет.

klntsky commented:
> При загрузке мобильного лейаута случается следующее:

Ты через devTools делал эмуляцию мобильного интерфейса?
По-моему это потому, что fastclick в этом случае ошибается в определении типа устройства.

limitedeternity commented:
> Это нужно потому, что на мобильных практически никто не пользуется списком локальных файлов.

Я пользуюсь переходом к строке по её номеру.

> Ты через devTools делал эмуляцию мобильного интерфейса?

Да

limitedeternity commented:
See #6

limitedeternity committed:
Remove polyfill.io

klntsky commented:
> Я пользуюсь переходом к строке по её номеру.

ОК. Думаю, мы можем просто сместить листинг файлов вниз, чтобы он не перекрывал кнопки.

#2

klntsky commented:
> Также, вероятно, README.md нужно обновить.

В другом бранче, please.

limitedeternity committed:
Edit config.content_dir

klntsky committed:
Merge pull request #6 from lex-js/server-fixes

Sync changes from server-fixes to rewrite-src

limitedeternity committed:
Replace Array.from

limitedeternity committed:
Promise & Fetch polyfill + fixes

klntsky committed:
added: native scrolling (src/scroll.js)
removed: additional lines at the end of file
removed: src/touch.js
changed: layout (reduced the use of absolute positioning)

TODO: update CSS styles related to content-browser

limitedeternity committed:
Merge branch 'rewrite-src' into native-scrolling

limitedeternity commented on src/content-browser.js:
```diff
@@ -88,14 +88,11 @@ module.exports = class ContentBrowser {
+      this.app.config.content_list_url + this.app.state.content_list.path
     );
+
+    return response.json();
```
```js
const json = await response.json();
return json;
```

klntsky replied:
Это одно и то же, и даже await здесь не обязателен.

Зачем вводить переменную, которую используешь один раз?

limitedeternity replied:
Нет, это не одно и то же. Метод json() у response возвращает `Promise<Object>`.
Нам нужен чисто объект, а не промис, поэтому нужно подождать, когда он резолвнется.

klntsky replied:
Вложенные друг в друга промисы - это то же самое, что и один промис.

Читаем https://javascript.info/promise-chaining#returning-promises

```
Normally, a value returned by a .then handler is immediately passed to the next handler. But there’s an exception.

If the returned value is a promise, then the further execution is suspended until it settles. After that, the result of that promise is given to the next .then handler.
```

### лирическое отступление

Вообще `Promise` ведёт себя как монада, когда функция, поданная в `.then` возвращает `Promise`.

И все законы монад соблюдаются в этом случае.

https://wiki.haskell.org/Monad_laws

Если на псевдоязыке, `response.json() >>= \json -> return json` - это то же, что `response.json() >>= return` (по правилу eta conversion).

А второй закон монад (right identity) говорит о том, что `m >>= return ≡ m`

Очень советую думать о `Promise` как о монаде, кстати - в случае, когда ситуация становится сложной, помогает.

limitedeternity replied:
Ух, жесть.
Проверил - всё так.

limitedeternity commented:
Может, сконвертировать @lex-js в организацию?

klntsky committed:
Get rid of `scroll-linked positioning effect` warning

klntsky committed:
Fixed: overlap between search field and search results.
Added: `blank_lines` user config option.

klntsky committed:
Restore original content-browser look before layout update

limitedeternity committed:
Merge pull request #10 from lex-js/native-scrolling

added: native scrolling (src/scroll.js)

klntsky committed:
Fix preloader for serverless usage

klntsky committed:
Actually use polyfills

klntsky committed:
use window.fetch

limitedeternity committed:
Merge pull request #11 from lex-js/use-polyfills

Use polyfills

limitedeternity committed:
Correct fetch.js & make HC launch test
